presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java

```diff
@@ -72,53 +80,89 @@ private RetryDriver()
     {
-        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier);
+        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier, retryExecutorService);
     }
 
     public RetryDriver withClassifier(Function<Exception, Exception> classifier)
     {
-        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier);
+        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier, retryExecutorService);
     }
 
+    public <V> ListenableFuture<V> runAsync(String callableName, Callable<ListenableFuture<V>> callable)
+    {
+        requireNonNull(callableName, "callableName is null");
+        requireNonNull(callable, "callable is null");
+        return runAsyncInternal(callableName, callable, new RetryStatus());
+    }
+
+    /**
+     * This is only for sync calls. For Callable that returns Future object, use runAsync instead.
+     */
     public <V> V run(String callableName, Callable<V> callable)
     {
         requireNonNull(callableName, "callableName is null");
         requireNonNull(callable, "callable is null");
+        return runSyncInternal(callableName, callable, new RetryStatus());
+    }
 
-        long startTime = System.nanoTime();
-        int attempt = 0;
-        while (true) {
-            attempt++;
+    private <V> ListenableFuture<V> runAsyncInternal(String callableName, Callable<ListenableFuture<V>> callable, RetryStatus retryStatus)
+    {
+        retryStatus.nextAttempt();
+        ListenableFuture<V> resultFuture;
+        try {
+            resultFuture = callable.call();
+        }
+        catch (InterruptedException ie) {
+            Thread.currentThread().interrupt();
+            throw propagate(ie);
+        }
+        catch (Exception e) {
+            resultFuture = immediateFailedFuture(e);
+        }
 
-            if (attempt > 1) {
-                retryRunnable.ifPresent(Runnable::run);
+        return Futures.catchingAsync(resultFuture, Exception.class, e -> {
+            if (e instanceof InterruptedException) {
+                Thread.currentThread().interrupt();
+                throw propagate(e);
             }
+            else if (retryStatus.shouldStopRetry(e)) {
+                throw propagate(e);
+            }
+            log.warn("Failed on executing %s with attempt %d, will perform async retry. Exception: %s", callableName, retryStatus.getAttempts(), e.toString());
+            return Futures.dereference(retryExecutorService.schedule(() -> runAsyncInternal(callableName, callable, retryStatus), retryStatus.getRetryDelayInMs(), MILLISECONDS));
+        });
+    }
 
+    private <V> V runSyncInternal(String callableName, Callable<V> callable, RetryStatus retryStatus)
```

Comment:
now you can inline this method, because it's not used by async

Reply:
Done